src/python/pants/task/simple_codegen_task.py

```diff
@@ -248,24 +276,39 @@ def _inject_synthetic_target(
     :param target: The target to inject a synthetic target for.
     :param target_workdir: The work directory containing the generated code for the target.
     :param fingerprint: The fingerprint to create the synthetic target
-           with to avoid re-fingerprinting
-    :param mark_transitive_invalidation_hash_dirty: Whether to walk the build graph to invalidate
-           dependees of the target.
+           with to avoid re-fingerprinting.
     """
     copied_attributes = {}
     for attribute in self._copy_target_attributes:
       copied_attributes[attribute] = getattr(target, attribute)
 
     target_workdir = self.synthetic_target_dir(target, target_workdir)
 
+    synthetic_extra_dependencies = self.synthetic_target_extra_dependencies(target, target_workdir)
+    synthetic_target_type = self.synthetic_target_type(target)
+
+    if hasattr(synthetic_target_type, 'export_specs'):
+      original_exports = self._original_export_specs(target)
+      extra_exports = self.synthetic_target_extra_exports(target, target_workdir)
+
+      extra_exports_not_in_extra_dependencies = set(extra_exports).difference(
+        set(synthetic_extra_dependencies))
+      if len(extra_exports_not_in_extra_dependencies) > 0:
+        raise self.MismatchedExtraExports(
+          'Extra synthetic exports included targets not in the extra dependencies: {}'
+            .format(extra_exports_not_in_extra_dependencies))
+      union = set(original_exports).union({e.address.spec for e in extra_exports})
+
+      copied_attributes['exports'] = sorted(union)
+
     sources = list(self.find_sources(target, target_workdir))
     if fingerprint:
       sources = self._create_sources_with_fingerprint(target_workdir, fingerprint, sources)
 
     synthetic_target = self.context.add_new_target(
       address=self._get_synthetic_address(target, target_workdir),
-      target_type=self.synthetic_target_type(target),
-      dependencies=self.synthetic_target_extra_dependencies(target, target_workdir),
+      target_type=synthetic_target_type,
+      dependencies=synthetic_extra_dependencies,
```

Comment:
As mentioned above, I think that if we force a topological sort here, you should be able to explicitly declare the dependencies on the target, which would avoid needing to inject them at all...? So this could be a complete list of the final dependencies of the target.

...or am I missing something? Not clear why we didn't do this ages ago.

Reply:
The only dependencies injected here are the extra ones. The regular dependencies are injected below.

Reply:
I understand that. I'm suggesting changing the model.

Reply:
The injections are setup the way they are to improve performance. When inject_target adds deps it does more work than is necessary for synthetic code gen. That does make me wonder whether the extra deps should be injected below too. The fact that we're doing both makes it extra confusing.

Reply:
Another approach I'd considered the last time I dug into this task was to collect all of the dependency changes and do the build graph edge modifications after all of the synthetic targets were created.